Feed environment data into RNG initializers

sipa opened the pull request:
This introduces a new `randomenv` module that queries varies non-cryptographic (and non-RNG) sources of entropy available on the system; things like user IDs, system configuration, time, statistics, CPUID data.

The idea is that these provide a fallback in scenarios where system entropy is somehow broken (note that if system entropy *fails* we will abort regardless; this is only meant to function as a last resort against undetected failure). It includes some data sources OpenSSL currently uses, and more.

The separation between random and randomenv is a bit arbitrary, but I felt that all this "non-essential" functionality deserved to be separated from the core random module.

fanquake committed:
doc: minor corrections in random.cpp

This should have been part of #17151.

sipa committed:
[MOVEONLY] Move perfmon data gathering to new randomenv module

sipa committed:
Seed randomness with process id / thread id / various clocks

This sort of data is also used by OpenSSL.

sipa committed:
Gather additional entropy from the environment

This based on code by Gregory Maxwell.

sipa committed:
Feed CPUID data into RNG

laanwj commented on src/randomenv.cpp:
```diff
@@ -0,0 +1,295 @@
+
+#if defined(__x86_64__) || defined(__amd64__) || defined(__i386__)
+    AddCPUID(hasher);
+#endif
```
To also have CPU features bits for other architectures (on Linux) you could include `getauxval(AT_HWCAP)` and `getauxval(AT_HWCAP2)`.

sipa replied:
Good idea, added!

laanwj commented:
Errors on MacOSX (looks like it doesn't have `environ` and `CLOCK_MONOTONIC`):
```
andomenv.cpp:177:19: error: use of undeclared identifier 'CLOCK_MONOTONIC'
    clock_gettime(CLOCK_MONOTONIC, &ts);
randomenv.cpp:222:64: error: use of undeclared identifier 'environ'
    hasher << &x << &RandAddStaticEnv << &malloc << &errno << &environ << addr;                                                
randomenv.cpp:282:9: error: use of undeclared identifier 'environ'
    if (environ) {
randomenv.cpp:283:28: error: use of undeclared identifier 'environ'
        for (size_t i = 0; environ[i]; ++i) {
randomenv.cpp:284:48: error: use of undeclared identifier 'environ'; did you mean 'union'?
            hasher.Write((const unsigned char*)environ[i], strlen(environ[i]));
                                               ^~~~~~~
                                               union
randomenv.cpp:284:48: error: expected expression
randomenv.cpp:284:67: error: use of undeclared identifier 'environ'
            hasher.Write((const unsigned char*)environ[i], strlen(environ[i]));
                                                                 ^
7 errors generated.
```

sipa committed:
Add information gathered through getauxval()

Suggested by Wladimir van der Laan.

sipa committed:
[MOVEONLY] Move cpuid code from random & sha256 to compat/cpuid

practicalswift commented:
Concept ACK: good idea

sipa commented:
I think I've addressed a compatibility/build issues. Can someone try this on OSX, and perhaps some BSD flavor?

RandyMcMillan commented:
@sipa, 
It runs....
https://github.com/RandyMcMillan/bitcoin/blob/201910_seedrandom/debug.log
https://github.com/RandyMcMillan/bitcoin/blob/201910_seedrandom/mojave10.14.6build.txt

RandyMcMillan commented:
https://travis-ci.org/RandyMcMillan/bitcoin/builds/603652597

RandyMcMillan commented:
travis-ci > https://travis-ci.org/RandyMcMillan/bitcoin/builds/603655421

promag reviewed (commented):
Concept ACK, nice commit sequence.

fanquake commented:
> Can someone try this on OSX, and perhaps some BSD flavor?

I get the following error when building eed8cfec452a538d644844fb96606e6bc24f47c8 on OpenBSD 6.6:
```
  CXX      libbitcoin_util_a-randomenv.o
In file included from randomenv.cpp:37:
/usr/include/netinet/ip.h:67:19: error: field has incomplete type 'struct in_addr'
        struct    in_addr ip_src, ip_dst; /* source and dest address */
                          ^
/usr/include/netinet/ip.h:67:11: note: forward declaration of 'in_addr'
        struct    in_addr ip_src, ip_dst; /* source and dest address */
                  ^
/usr/include/netinet/ip.h:67:27: error: field has incomplete type 'struct in_addr'
        struct    in_addr ip_src, ip_dst; /* source and dest address */
                                  ^
/usr/include/netinet/ip.h:67:11: note: forward declaration of 'in_addr'
        struct    in_addr ip_src, ip_dst; /* source and dest address */
                  ^
/usr/include/netinet/ip.h:181:19: error: field has incomplete type 'struct in_addr'
                        struct in_addr ipt_addr;
                                       ^
/usr/include/netinet/ip.h:67:11: note: forward declaration of 'in_addr'
        struct    in_addr ip_src, ip_dst; /* source and dest address */
                  ^
randomenv.cpp:101:57: error: unknown type name 'sockaddr_in'; did you mean 'sockaddr'?
        hasher.Write((const unsigned char*)addr, sizeof(sockaddr_in));
                                                        ^~~~~~~~~~~
                                                        sockaddr
/usr/include/sys/socket.h:207:8: note: 'sockaddr' declared here
struct sockaddr {
       ^
randomenv.cpp:104:57: error: unknown type name 'sockaddr_in6'; did you mean 'sockaddr'?
        hasher.Write((const unsigned char*)addr, sizeof(sockaddr_in6));
                                                        ^~~~~~~~~~~~
                                                        sockaddr
/usr/include/sys/socket.h:207:8: note: 'sockaddr' declared here
struct sockaddr {
       ^
5 errors generated.
```

laanwj commented:
Log on FreeBSD 12.0-RELEASE-p10 (same complaint, about `sockaddr`)
<details>
```
Making all in src
gmake[1]: Entering directory '/usr/home/user/src/bitcoin/src'
gmake[2]: Entering directory '/usr/home/user/src/bitcoin/src'
  CXX      libbitcoin_util_a-randomenv.o
In file included from randomenv.cpp:37:
/usr/include/netinet/ip.h:71:17: error: field has incomplete type 'struct in_addr'
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
                        ^
/usr/include/netinet/ip.h:71:9: note: forward declaration of 'in_addr'
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
                ^
/usr/include/netinet/ip.h:71:24: error: field has incomplete type 'struct in_addr'
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
                               ^
/usr/include/netinet/ip.h:71:9: note: forward declaration of 'in_addr'
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
                ^
/usr/include/netinet/ip.h:188:19: error: field has incomplete type 'struct in_addr'
                        struct in_addr ipt_addr;
                                       ^
/usr/include/netinet/ip.h:71:9: note: forward declaration of 'in_addr'
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
                ^
/usr/include/netinet/ip.h:223:17: error: field has incomplete type 'struct in_addr'
        struct  in_addr ippseudo_src;   /* source internet address */
                        ^
/usr/include/netinet/ip.h:71:9: note: forward declaration of 'in_addr'
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
                ^
/usr/include/netinet/ip.h:224:17: error: field has incomplete type 'struct in_addr'
        struct  in_addr ippseudo_dst;   /* destination internet address */
                        ^
/usr/include/netinet/ip.h:71:9: note: forward declaration of 'in_addr'
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
                ^
randomenv.cpp:101:57: error: unknown type name 'sockaddr_in'; did you mean 'sockaddr'?
        hasher.Write((const unsigned char*)addr, sizeof(sockaddr_in));
                                                        ^~~~~~~~~~~
                                                        sockaddr
/usr/include/sys/socket.h:328:8: note: 'sockaddr' declared here
struct sockaddr {
       ^
randomenv.cpp:104:57: error: unknown type name 'sockaddr_in6'; did you mean 'sockaddr'?
        hasher.Write((const unsigned char*)addr, sizeof(sockaddr_in6));
                                                        ^~~~~~~~~~~~
                                                        sockaddr
/usr/include/sys/socket.h:328:8: note: 'sockaddr' declared here
struct sockaddr {
       ^
7 errors generated.
gmake[2]: *** [Makefile:8866: libbitcoin_util_a-randomenv.o] Error 1
gmake[2]: Leaving directory '/usr/home/user/src/bitcoin/src'
gmake[1]: *** [Makefile:14135: all-recursive] Error 1
gmake[1]: Leaving directory '/usr/home/user/src/bitcoin/src'
gmake: *** [Makefile:774: all-recursive] Error 1
```
</details>

MarcoFalke commented:
Building freeBSD here: https://cirrus-ci.com/task/5690364417015808

MarcoFalke commented:
Does the comment in `random.h` need to be updated?

fanquake commented:
> Does the comment in random.h need to be updated?

If that is updated, you could also cherry-pick https://github.com/bitcoin/bitcoin/pull/17265/commits/770cd9640e8b38d3456527bdf3d84b589cbd3418 out of #17265 which contains some related corrections.

sipa committed:
Use sysctl for seeding on MacOS/BSD

MarcoFalke commented:
New freeBSD build: https://cirrus-ci.com/task/5120748003262464

MarcoFalke commented:
New freeBSD build failure: https://cirrus-ci.com/task/4712584309112832?command=make#L761

<details><summary>log</summary>

```
/usr/local/bin/ccache c++ -std=c++11 -DHAVE_CONFIG_H -I. -I../src/config   -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -I. -I/tmp/cirrus-ci-build/db4/include -DBOOST_SP_USE_STD_ATOMIC -DBOOST_AC_USE_STD_ATOMIC -pthread -I/usr/local/include -I./leveldb/include -I./leveldb/helpers/memenv -I/usr/local/include  -I./secp256k1/include -I./univalue/include -Qunused-arguments  -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS  -Wstack-protector -fstack-protector-all -Wall -Wextra -Wformat -Wvla -Wswitch -Wformat-security -Wthread-safety-analysis -Wrange-loop-analysis -Wredundant-decls -Wno-unused-parameter -Wno-self-assign -Wno-unused-local-typedef -Wno-deprecated-register -Wno-implicit-fallthrough    -fPIE -g -O2 -c -o libbitcoin_util_a-randomenv.o `test -f 'randomenv.cpp' || echo './'`randomenv.cpp
In file included from randomenv.cpp:29:
/usr/include/netinet/ip.h:53:2: error: unknown type name 'u_char'; did you mean 'char'?
        u_char  ip_hl:4,                /* header length */
        ^
/usr/include/netinet/ip.h:60:2: error: unknown type name 'u_char'; did you mean 'char'?
        u_char  ip_tos;                 /* type of service */
        ^
/usr/include/netinet/ip.h:61:2: error: unknown type name 'u_short'; did you mean 'short'?
        u_short ip_len;                 /* total length */
        ^
/usr/include/netinet/ip.h:62:2: error: unknown type name 'u_short'; did you mean 'short'?
        u_short ip_id;                  /* identification */
        ^
/usr/include/netinet/ip.h:63:2: error: unknown type name 'u_short'; did you mean 'short'?
        u_short ip_off;                 /* fragment offset field */
        ^
/usr/include/netinet/ip.h:68:2: error: unknown type name 'u_char'; did you mean 'char'?
        u_char  ip_ttl;                 /* time to live */
        ^
/usr/include/netinet/ip.h:69:2: error: unknown type name 'u_char'; did you mean 'char'?
        u_char  ip_p;                   /* protocol */
        ^
/usr/include/netinet/ip.h:70:2: error: unknown type name 'u_short'; did you mean 'short'?
        u_short ip_sum;                 /* checksum */
        ^
/usr/include/netinet/ip.h:71:17: error: field has incomplete type 'struct in_addr'
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
                        ^
/usr/include/netinet/ip.h:71:9: note: forward declaration of 'in_addr'
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
                ^
/usr/include/netinet/ip.h:71:24: error: field has incomplete type 'struct in_addr'
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
                               ^
/usr/include/netinet/ip.h:71:9: note: forward declaration of 'in_addr'
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
                ^
/usr/include/netinet/ip.h:174:2: error: unknown type name 'u_char'; did you mean 'char'?
        u_char  ipt_code;               /* IPOPT_TS */
        ^
/usr/include/netinet/ip.h:175:2: error: unknown type name 'u_char'; did you mean 'char'?
        u_char  ipt_len;                /* size of structure (variable) */
        ^
/usr/include/netinet/ip.h:176:2: error: unknown type name 'u_char'; did you mean 'char'?
        u_char  ipt_ptr;                /* index of current entry */
        ^
/usr/include/netinet/ip.h:178:2: error: unknown type name 'u_char'; did you mean 'char'?
        u_char  ipt_flg:4,              /* flags, see below */
        ^
/usr/include/netinet/ip.h:188:19: error: field has incomplete type 'struct in_addr'
                        struct in_addr ipt_addr;
                                       ^
/usr/include/netinet/ip.h:71:9: note: forward declaration of 'in_addr'
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
                ^
/usr/include/netinet/ip.h:223:17: error: field has incomplete type 'struct in_addr'
        struct  in_addr ippseudo_src;   /* source internet address */
                        ^
/usr/include/netinet/ip.h:71:9: note: forward declaration of 'in_addr'
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
                ^
/usr/include/netinet/ip.h:224:17: error: field has incomplete type 'struct in_addr'
        struct  in_addr ippseudo_dst;   /* destination internet address */
                        ^
/usr/include/netinet/ip.h:71:9: note: forward declaration of 'in_addr'
        struct  in_addr ip_src,ip_dst;  /* source and dest address */
                ^
/usr/include/netinet/ip.h:225:2: error: unknown type name 'u_char'; did you mean 'char'?
        u_char          ippseudo_pad;   /* pad, must be zero */
        ^
/usr/include/netinet/ip.h:226:2: error: unknown type name 'u_char'; did you mean 'char'?
        u_char          ippseudo_p;     /* protocol */
        ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
gmake[2]: *** [Makefile:8866: libbitcoin_util_a-randomenv.o] Error 1
gmake[2]: Leaving directory '/tmp/cirrus-ci-build/src'
gmake[1]: *** [Makefile:14135: install-recursive] Error 1
gmake[1]: Leaving directory '/tmp/cirrus-ci-build/src'
gmake: *** [Makefile:774: install-recursive] Error 1
Exit status: 2
```

</details>

MarcoFalke commented:
new build :pray: https://cirrus-ci.com/task/6286311933083648

sipa commented:
@MarcoFalke Hmm, "checking whether sysctlbyname is declared... no". Does FreeBSD not have `sysctlbyname`?

fanquake commented:
> MarcoFalke Hmm, "checking whether sysctlbyname is declared... no". Does FreeBSD not have sysctlbyname?

It does, looks like you'll just need to include `sys/types.h` as well. See the [sysctlbyname docs](http://nixdoc.net/man-pages/FreeBSD/man3/sysctlbyname.3.html).

After including `sys/types.h`I've tested using `12.0-RELEASE-p6` and see `sysctlbyname is declared... yes`.

MarcoFalke commented:
New build: https://cirrus-ci.com/task/5389706841292800

MarcoFalke commented:
For obviously broken cases (e.g. the random number generator returns a constant) it is possible to detect it with automated tools that apply heuristics such as those described in https://web.archive.org/web/20110602194701/http://users.eecs.northwestern.edu/~nickle/310/2010/headRuns.pdf . However, these tools can only hint at potential problems and never certify correctness. See https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-22r1a.pdf for an example of such a tool. That tool only works for long sequences, so it doesn't seem applicable to the sources added here... Not to mention the randomness sources added in this pull request are non-cryptographic, so any slightly sane tool would immediately yell at them.

Sjors reviewed (commented):
Code review ACK on the move-only stuff up to  35cfbf7 `Move cpuid code from random & sha256 to compat/cpuid`. Concept ACK on the followup commits that add more sources (5a078e2);  looks sane, but not sure how to review.

I'm able to run the test suite on macOS 10.15 (using `MACOSX_DEPLOYMENT_TARGET=10.14` to avoid https://github.com/bitcoin-core/secp256k1/issues/674).

sipa commented on src/crypto/sha256.cpp:
```diff
@@ -8,15 +8,14 @@
+#include <compat/cpuid.h>
+
 #if defined(__x86_64__) || defined(__amd64__) || defined(__i386__)
-#if defined(USE_ASM)
```
Fixed.

sipa committed:
Run background seeding periodically instead of unpredictably

* Instead of calling RandAddSeedSleep anytime the scheduler goes
  idle, call its replacement (RandAddSeedPeriodic) just once per
  minute. This has better guarantees of actually being run, and
  helps limit how frequently the dynamic env data is gathered.
* Since this code runs once per minute regardless now, we no
  longer need to keep track of the last time strengthening was
  run; just do it always.
* Make strengthening time context dependent (100 ms at startup,
  10 ms once per minute afterwards).

sipa committed:
Use thread-safe atomic in perfmon seeder

Also switch to chrono based types.

sipa commented:
A bunch of updates over the past days:
* Cherry-picked https://github.com/bitcoin/bitcoin/commit/770cd9640e8b38d3456527bdf3d84b589cbd3418
* Added a moveonly commit to merge the cpuid code in random and crypto/sha256, making it reusable for randomenv.
* Added a commit to use `getauxval` on Linux.
* Added a commit to use `sysctlbyname` on BSD/MacOS.
* Added a commit to make sure dynamic env seeders do not run more than once per minute, and added a thread to make sure they do get run occasionally.
* Added a commit to fix a bug in the perfmon seeder (now that it can be called from multiple threads, it may need an atomic to track the previous run time).

gmaxwell commented:
A reminder that this sort of thing is important:  Some ryzen cpus rng returns a constant (and success), and in default configs linux relies on it for many purposes (I think, /dev/random in current default kernel configs mostly manages to dodge the over-reliance, but many other things do-- so it's a narrow miss.)

laanwj commented:
Travis error on mingw:
```
randomenv.cpp: In function ‘void {anonymous}::RandAddSeedPerfmon(CSHA512&)’:
randomenv.cpp:63:33: error: variable ‘std::atomic<long long int> last_perfmon’ has initializer but incomplete type
     static std::atomic<int64_t> last_perfmon{0};
                                 ^~~~~~~~~~~~
```

MarcoFalke commented on src/randomenv.cpp:
```diff
@@ -60,10 +61,11 @@ void RandAddSeedPerfmon(CSHA512& hasher)
-    nLastPerfmon = GetTime();
+    static std::atomic<int64_t> last_perfmon{0};
+    int64_t last_time = last_perfmon.load();
+    int64_t current_time = GetTime();
```
in commit ad09ad707e0ed6e3a537e1a204843792e5d06108:

`GetTime` is deprecated (see https://dev.visucore.com/bitcoin/doxygen/time_8cpp.html#a350f99e2a13df31f2afbd7f80ab21a5e). I guess you wanted to use `GetSystemTimeInSeconds`

MarcoFalke replied:
`GetTime<std::chrono::seconds>()` can't hurt here either, as far as I can tell

sipa replied:
Fixed. Also done elsewhere in the previous commit.

fanquake commented on src/randomenv.cpp:
```diff
@@ -0,0 +1,389 @@
+#ifdef __linux__
+    AddFile(hasher, "/proc/diskstats");
+    AddFile(hasher, "/proc/vmstat");
+    AddFile(hasher, "/proc/schedstat");
```
I think this should be `/proc/self/schedstat`

sipa replied:
It seems both exist.

fanquake commented on src/randomenv.cpp:
```diff
@@ -0,0 +1,389 @@
+#endif
+
+    AddSysctl(hasher, "vm.loadavg");
+    AddSysctl(hasher, "vm.swapusage");
```
Note `vm.swapusage` seems to be macOS only. I can't find it on OpenBSD or FreeBSD.

sipa replied:
That's ok; I combined lists of sysctls listed for various platforms. Not all will exist on each.

fanquake commented on src/randomenv.cpp:
```diff
@@ -0,0 +1,389 @@
+    AddFile(hasher, "/etc/resolv.conf");
+    AddFile(hasher, "/etc/timezone");
+    AddFile(hasher, "/etc/localtime");
+    AddFile(hasher, "/etc/hostconfig");
```
I couldn't find `/etc/hostconfig` on any of the platforms I tested (macOS, *BSD, Linux). Seems it used to exist on at least macOS, but was removed in `10.10`.

sipa replied:
Ok, removing.

fanquake commented on src/randomenv.cpp:
```diff
@@ -0,0 +1,389 @@
+    hasher << mach_absolute_time();
+#  else
+    struct timespec ts;
+    clock_gettime(CLOCK_MONOTONIC, &ts);
```
It looks like upstream OpenSSL uses `CLOCK_REALTIME` when generating a timestamp. Could you explain why `CLOCK_MONOTONIC` is the better choice for us? I assume it's TODO with it being less manipulable and kernel compatibility?

sipa replied:
I'm expanding this piece of code to just use all clocks available.

fanquake commented on src/randomenv.cpp:
```diff
@@ -0,0 +1,389 @@
+    hasher << ftime;
+#else
+#  ifdef __MACH__
+    hasher << mach_absolute_time();
```
There's not a heap of documentation on  `mach_absolute_time` and I can't find any calls to it in recent OpenSSL, however looks like it's available for macOS versions > 10.9. `/usr/include/mach/mach_time.h`.

sipa replied:
Yeah, I tried to find an equivalent to CLOCK_MONOTONIC on MacOS, and found this. It doesn't exactly match what OpenSSL did.

fanquake reviewed (commented):
Concept ACK

`sysctlbyname` isn't being picked up on OpenBSD (works on macOS and FreeBSD). Will investigate.

I've pushed some WIP review notes up to https://github.com/fanquake/bitcoin/tree/17270_annotated.

fanquake commented:
> sysctlbyname isn't being picked up on OpenBSD (works on macOS and FreeBSD). Will investigate.

Looks like it doesn't actually exist on OpenBSD. It only has [`sysctl`](https://man.openbsd.org/OpenBSD-6.6/sysctl.2). On OpenBSD 6.6:
```bash
bazinga# cat sysctl.h | grep sysctlbyname
bazinga# cat sysctl.h | grep "sysctl("
 * pairs to be used by sysctl(1) in manipulating the subsystem.
 * aggregated into an array in debug_sysctl(), so that it can
int kern_sysctl(int *, u_int, void *, size_t *, void *, size_t,
int hw_sysctl(int *, u_int, void *, size_t *, void *, size_t,
< trimmed >
int mpls_sysctl(int *, u_int, void *, size_t *, void *, size_t);
int	sysctl(const int *, u_int, void *, size_t *, void *, size_t);
```

sipa commented:
Addressed the comments above (AT_RANDOM gives a pointer, removed /etc/hostconfig, added a comment that not all sysctls are available on every system).

Also added:
* Added more clocks (they're gathered once a minute, some redundancy won't hurt).
* Improved the CPUID code (also adds extended leaves now)

sipa commented:
@fanquake Ok, annoying. Perhaps that means we should use sysctl instead of sysctlbyname.

sipa commented on src/randomenv.cpp:
```diff
@@ -0,0 +1,389 @@
+    hasher << getauxval(AT_HWCAP2);
+#  endif
+#  ifdef AT_RANDOM
+    hasher << getauxval(AT_RANDOM);
```
Fixed.

laanwj commented:
Changes now look good to me. Tested on Linux RISC-V 64 bit.

sipa commented:
Replaced the use of `sysctlbyname` by `sysctl`; it's less extensive and needs more code, but likely supports more systems.

laanwj commented on src/randomenv.cpp:
```diff
@@ -0,0 +1,507 @@
+template<int... S>
+void AddSysctl(CSHA512& hasher)
+{
+#if HAVE_DECL_SYSCTL
```
Shouldn't this ifdef be outside the function instead of outside it? I mean, I don't think it should be possible to call `AddSysctl` on systems without sysctl?

sipa replied:
Fixed.

laanwj commented on src/randomenv.cpp:
```diff
@@ -0,0 +1,507 @@
+    // From https://gist.github.com/jbenet/1087739
+    clock_serv_t cclock;
+    mach_timespec_t mts;
+    host_get_clock_service(mach_host_self(), CALENDAR_CLOCK, &cclock);
```
could creating the service fail?

sipa replied:
Fixed.

laanwj commented:
it's pretty hard to verify that this is doing anything, at runtime; it would be too much hassle to keep track of everything that is hashed, but it would be nice to have (optionally enabled) debug log messages like;
`N bytes of static environment data added to random state`
`N bytes of dynamic environment data added to random state`

sipa committed:
Report amount of data gathered from environment

narula commented on src/randomenv.cpp:
```diff
@@ -81,6 +99,58 @@ CSHA512& operator<<(CSHA512& hasher, const T& data) {
+            n = read(f, fbuf, sizeof(fbuf));
+            if (n > 0) hasher.Write(fbuf, n);
+            /* not bothering with EINTR handling. */
+        } while (n == sizeof(fbuf));
```
Might you consider only reading up to a certain number of bytes? I'm not sure you're guaranteed that some of these files won't be enormous.

sipa replied:
Done.

sipa commented:
@laanwj I've added a commit that logs the number of bytes gathered.

madars commented on src/randomenv.cpp:
```diff
@@ -0,0 +1,515 @@
+/** Helper to easily feed data into a CSHA512.
+ *
+ * Note that this does not serialize the passed object (like stream.h's << operators do).
+ * Its raw memory representation is used directly.
```
Should we make this comment explicitly enforced, e.g. having `static_assert(!std::is_same<T, char*>::value, "calling hasher.operator<< on char* probably won't do what you want");` or similar?

laanwj replied:
it's only defined in this cpp file, where the hasher is only used for one thing, so the scope for potential accidental misuse is very little (and who knows, maybe you'd *want* to hash a pointer here, if it's random enough due to ASLR)

sipa replied:
There are plenty of internal pointers being fed to the RNG here (stack locations, standard library locations, ...).

madars replied:
They don't appear to be `char*` pointers (e.g. `environ` is `char**`, result of a temp malloc below is `void*`, etc) but potential for misuse is a judgement call.

sipa replied:
Ah, good point. Let's see if that compiles.

sipa replied:
Done. I added `std::decay<T>` so that it also catches const char* and char arrays.

laanwj commented:
Tested on a few different platforms;

FreeBSD x86_64:
```
2019-11-10T13:02:15Z Feeding 66012 of dynamic environment data into RNG
2019-11-10T13:02:15Z Feeding 10350 bytes of static environment data into RNG
```
Linux x86_64 VM:
```
2019-11-10T15:33:02Z Feeding 16094 of dynamic environment data into RNG
2019-11-10T15:33:02Z Feeding 14944 bytes of static environment data into RNG
```
Linux aarch64:
```
2019-11-10T16:18:13Z Feeding 11043 of dynamic environment data into RNG
2019-11-10T16:18:13Z Feeding 7073 bytes of static environment data into RNG
```
Linux arm32:
```
2019-11-10T15:52:06Z Feeding 10921 of dynamic environment data into RNG
2019-11-10T15:52:06Z Feeding 8906 bytes of static environment data into RNG
```
Linux riscv64:
```
2019-11-10T16:50:38Z Feeding 10389 of dynamic environment data into RNG
2019-11-10T16:50:38Z Feeding 6833 bytes of static environment data into RNG
```


New compile warning (gcc Linux):
```
In file included from randomenv.cpp:50:
/usr/include/sys/sysctl.h:21:2: warning: #warning "The <sys/sysctl.h> header is deprecated and will be removed." [-Wcpp]
   21 | #warning "The <sys/sysctl.h> header is deprecated and will be removed."
      |  ^~~~~~~
```
Looks this was a recent libc/kernel change (http://patches-tcwg.linaro.org/patch/19443/). I don't think the intent is to use sysctl on Linux (it already reads `/proc` directly, which is what they suggest instead), so we might want to prevent its detection there somehow.

sipa commented:
Added a commit to always avoids sysctl on Linux.

laanwj commented on src/randomenv.cpp:
```diff
@@ -46,6 +46,7 @@
 #if HAVE_DECL_GETIFADDRS
 #include <ifaddrs.h>
 #endif
+#ifndef __linux__
```
Thanks!
Though, I'd slightly prefer to handle this in the build system, so that the symbol `HAVE_DECL_SYSCTL` doesn't get set in the first place in Linux. This avoids having to handle it on a case by case basis on all caller sites.

sipa replied:
Good point, fixed.

TheBlueMatt commented on src/random.cpp:
```diff
@@ -585,6 +603,9 @@ static void ProcRand(unsigned char* out, int num, RNGLevel level)
     case RNGLevel::SLEEP:
         SeedSleep(hasher, rng);
         break;
+    case RNGLevel::PERIODIC:
```
I don't get the point of this commit.
You're avoiding doing something that sleeps 1ms but runs in a tight loop for 10ms.
The real meat of the function gets called on a timer every minute, but also my be called at other times, but only runs if its been 10 minutes since the last run.

Maybe we should just skip the SeedInfrequent call in SeedSleep and then we can drop all the once-per-minute checking in this file wholesale and use the scheduler?

sipa replied:
I think you're right, there is fairly little advantage to having both RandAddSeedSleep and RandAddPeriodic. I've redone this PR to instead just have a RandAddPeriodic, which indeed simplifies a bunch of things including not keeping track of when the last run of the expensive stuff was (there is a still a counter in the perfmon data for windows, which only runs once every 10 minutes instead of every minute).

TheBlueMatt reviewed (commented):
Awesome work! One real comment, but otherwise looks good.

sipa commented on src/randomenv.cpp:
```diff
@@ -0,0 +1,522 @@
+
+void AddStat(CSHA512& hasher, const struct stat& sb)
+{
+    hasher << sb.st_dev << sb.st_ino << sb.st_mode << sb.st_nlink << sb.st_uid << sb.st_gid << sb.st_size << sb.st_blksize << sb.st_blocks << sb.st_atime << sb.st_mtime << sb.st_ctime;
```
One reason to writing the fields directly is because these structures may have implementation-defined other fields, which are perhaps left uninitialized. Feeding them into the hasher would then trip up valgrind & co. Perhaps this is not a real concern and we should just feed the struct in until someone notices a problem?

sipa replied:
Ok, fixed: I'm zero initializing all the structs, and then just feeding them directly into the hasher.

TheBlueMatt replied:
I can't say I feel strongly. I see reasons to do it both ways, and as long as valgrind doesn't complain and we hash in the higher-entropy bits, it doesn't matter either way. Just figured I'd flag it since you appeared to be very careful about it in all places but here.

TheBlueMatt commented:
utACK d1c02775aa74a0610809ac54bb241ddad61d2d8c. Certainly no longer measuring the time elapsed between a 1ms sleep (which got removed in the latest change) is a fair tradeoff for adding about 2 million other actually-higher-entropy bits :).

fanquake commented:
I've been testing d1c02775aa74a0610809ac54bb241ddad61d2d8c across various operating systems. Few logs for anyone who's interested. Will post a final review shortly.

#### macOS 10.14.6
```bash
2019-11-13T18:24:03Z Feeding 1066350 bytes of environment data into RNG
2019-11-13T18:25:03Z Feeding 216 bytes of dynamic environment data into RNG
2019-11-13T18:26:03Z Feeding 216 bytes of dynamic environment data into RNG
```

#### OpenBSD 6.6
```bash
2019-11-13T19:21:17Z Feeding 9334 bytes of environment data into RNG
2019-11-13T19:22:18Z Feeding 504 bytes of dynamic environment data into RNG
2019-11-13T19:23:18Z Feeding 504 bytes of dynamic environment data into RNG
```

#### FreeBSD 12
```bash
2019-11-13T21:04:50Z Feeding 74253 bytes of environment data into RNG
2019-11-13T21:05:51Z Feeding 66068 bytes of dynamic environment data into RNG
2019-11-13T21:06:51Z Feeding 66068 bytes of dynamic environment data into RNG
```

#### Windows 10 WSL
```bash
2019-11-13T23:05:00Z Feeding 435111 bytes of environment data into RNG
2019-11-13T23:06:00Z Feeding 48 bytes of dynamic environment data into RNG
2019-11-13T23:07:00Z Feeding 48 bytes of dynamic environment data into RNG
```

#### Windows 10 MSVC
```bash
2019-11-13T23:16:08Z Feeding 378764 bytes of environment data into RNG
2019-11-13T23:17:08Z Feeding 48 bytes of dynamic environment data into RNG
2019-11-13T23:18:08Z Feeding 48 bytes of dynamic environment data into RNG
```

#### NetBSD 8.1
```bash
2019-11-13T23:50:05Z Feeding 35503 bytes of environment data into RNG
2019-11-13T23:51:05Z Feeding 28268 bytes of dynamic environment data into RNG
2019-11-13T23:52:05Z Feeding 28268 bytes of dynamic environment data into RNG
```

#### Alpine
```bash
2019-11-14T00:45:33Z Feeding 23631 bytes of environment data into RNG
2019-11-14T00:46:33Z Feeding 14196 bytes of dynamic environment data into RNG
2019-11-14T00:47:33Z Feeding 14206 bytes of dynamic environment data into RNG
2019-11-14T00:48:33Z Feeding 14208 bytes of dynamic environment data into RNG
2019-11-14T00:49:33Z Feeding 14211 bytes of dynamic environment data into RNG
```

#### Debian 10
```bash
2019-11-14T00:21:13Z Feeding 23714 bytes of environment data into RNG
2019-11-14T00:22:13Z Feeding 13980 bytes of dynamic environment data into RNG
2019-11-14T00:23:13Z Feeding 13983 bytes of dynamic environment data into RNG
2019-11-14T00:24:13Z Feeding 13991 bytes of dynamic environment data into RNG
2019-11-14T00:25:13Z Feeding 13988 bytes of dynamic environment data into RNG
2019-11-14T00:26:13Z Feeding 13993 bytes of dynamic environment data into RNG
2019-11-14T00:27:13Z Feeding 14024 bytes of dynamic environment data into RNG
```

fanquake commented on src/randomenv.cpp:
```diff
@@ -0,0 +1,508 @@
+    GetSystemTimeAsFileTime(&ftime);
+    hasher << ftime;
+#else
+#  ifndef __MACH__
```
Using `__MACH__` here and above is unusual for a our codebase, we use `MAC_OSX` throughout. Also `__MACH__` by itself isn't necessarily macOS, looks like if you were to use it you'd want to couple it with `__APPLE__`.

fanquake commented on src/randomenv.cpp:
```diff
@@ -0,0 +1,508 @@
+    hasher << ts;
+#    endif
+#  else
+    // On MacOS use mach_absolute_time (number of CPU ticks since boot) as a replacement for CLOCK_MONOTONIC,
```
After #16392 (bumping our SDK and minimum required macOS to 10.12), `clock_gettime`, `CLOCK_REALTIME` and `CLOCK_MONOTONIC` will be available to use on macOS. Not necessarily anything that needs to change here now, but the potential to simplify this later.

fanquake reviewed (commented):
I think the changes here should be sufficient that we can remove our OpenSSL usage (#17265). This is now likely doing above and beyond what (depends) OpenSSL was doing. I prefer the recent refactor to `RandAddPeriodic` and removing the RandAdd call from scheduler.cpp. I've done some sanity checking across various operating systems.

laanwj commented:
ACK d1c02775aa74a0610809ac54bb241ddad61d2d8c

I agree with @fanquake. I think this is good for merge. It has been tested on a wide range of platforms, and unarguably works.

Additional bits of entropy for specific arch and OS can be added later.

hebasto commented on src/randomenv.cpp:
```diff
@@ -0,0 +1,508 @@
+#include <sys/auxv.h>
+#endif
+
+//! Necessary on some platforms
+extern char** environ;
```
@sipa
While compiling natively [on Windows with MSVC](https://github.com/bitcoin/bitcoin/pull/25797#issuecomment-1215983238), receiving [`C4273`](https://docs.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-1-c4273) warnings:
```
  randomenv.cpp
C:\Users\hebasto\bitcoin\src\randomenv.cpp(60,22): warning C4273: '__p__environ': inconsistent dll linkage [C:\Users\hebasto\bitcoin\build\src\util\bitcoin_util.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\ucrt\stdlib.h(1158,29): message : see previous definition of '__p__environ' [C:\Users\hebasto\bitcoin\build\src\util\bitcoin_util.
vcxproj]
```

```
  randomenv.cpp
C:\Users\hebasto\bitcoin\src\randomenv.cpp(60,22): warning C4273: '__p__environ': inconsistent dll linkage [C:\Users\hebasto\bitcoin\build\src\kernel\bitcoinkernel.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.19041.0\ucrt\stdlib.h(1158,29): message : see previous definition of '__p__environ' [C:\Users\hebasto\bitcoin\build\src\kernel\bitcoinkern
el.vcxproj]
```

Considering https://docs.microsoft.com/en-us/cpp/c-runtime-library/environ-wenviron, is this code correct for MSVC?

hebasto replied:
nm, no issues with the latest [build](https://cirrus-ci.com/task/6359296752484352).
